### DISCUSSION TITLE
Documentation

### DISCUSSION THREAD

**Discussion by paulhamsh** [hathach/tinyusb#1666] [General]
Hi all
I am very impressed with TinyUSB and want to understand it better - is there any documentation of the files and function calls, and how to create a new host class?
I have seen the .org site but can’t find any details.

If not, I am interested in working on that - I’d need to start something for my own purposes so I can understand it, so happy to work on a community effort if that was of use.
Some in-code doxygen and some other pages describing how each layer works, the functions calls and so on.

What does everyone think, especially @hathach ?

**Comment by paulhamsh** [comment]
In case it helps I wrote this for a TLSF memory allocator - to help me work out why a NimBLE function didn’t seem to work on an ESP32-C3 but did on a plain old ESP32. It wasn’t BLE - it was my bad C++ pointer handling and interaction with the TLSF allocator.

https://github.com/paulhamsh/TLSF-Memory-Allocator-Overview

I like understanding code :-)

**Comment by hathach** [comment]
yeah, documentation is thing I would love to do more but couldn't have time (plus I am not that good with technical writing). APIs are install evolving, the text docs only cover basic concept to get started. For now, reading examples and source code is what I could suggest.
